sys/net/gnrc/link_layer/lwmac/lwmac.c

```diff
@@ -0,0 +1,745 @@
+        break;
+
+    /* Receiving incoming data */
+    case RECEIVING:
+        rtt_handler(LWMAC_EVENT_RTT_PAUSE); /* No duty cycling while TXing */
+        _set_netdev_state(&lwmac, NETOPT_STATE_IDLE);  /* Power up netdev */
+        break;
+
+    case STOPPED:
+        _set_netdev_state(&lwmac, NETOPT_STATE_OFF);
+        break;
+
+    /*********************** Control states ***********************************/
+    case START:
+        rtt_handler(LWMAC_EVENT_RTT_START);
+        lwmac_set_state(LISTENING);
+        break;
+
+    case STOP:
+        rtt_handler(LWMAC_EVENT_RTT_STOP);
+        lwmac_set_state(STOPPED);
+        break;
+
+    case RESET:
+        LOG_WARNING("Reset not yet implemented\n");
+        lwmac_set_state(STOP);
+        lwmac_set_state(START);
+        break;
+
+    /**************************************************************************/
+    default:
+        LOG_DEBUG("No actions for entering state %u\n", newstate);
+        return;
+    }
+
+    lwmac_schedule_update();
+}
+
+/******************************************************************************/
+
+/* Main state machine. Call whenever something happens */
+// TODO: Don't use global variables
+bool lwmac_update(void)
+{
+    lwmac.needs_rescheduling = false;
+
+    switch(lwmac.state)
+    {
+    case SLEEPING:
+
+        /* If a packet is scheduled, no other (possible earlier) packet can be
+         * sent before the first one is handled, even no broadcast
+         */
+        if( !lwmac_timeout_is_running(&lwmac, TIMEOUT_WAIT_FOR_DEST_WAKEUP) ) {
+
+            /* Check if there are broadcasts to send and transmit immediately */
+            if(packet_queue_length(&(lwmac.tx.neighbours[0].queue)) > 0) {
+                lwmac.tx.current_neighbour = &(lwmac.tx.neighbours[0]);
+                lwmac_set_state(TRANSMITTING);
+                break;
+            }
+
+            lwmac_tx_neighbour_t* neighbour = _next_tx_neighbour(&lwmac);
+
+            if(neighbour != NULL) {
+                /* Offset in microseconds when the earliest (phase) destination
+                 * node wakes up that we have packets for. */
+                int time_until_tx = RTT_TICKS_TO_US(_ticks_until_phase(neighbour->phase));
+
+                /* If there's not enough time to prepare a WR to catch the phase
+                 * postpone to next interval */
+                if (time_until_tx < LWMAC_WR_PREPARATION_US) {
+                    time_until_tx += LWMAC_WAKEUP_INTERVAL_US;
+                }
+
+                time_until_tx -= LWMAC_WR_PREPARATION_US;
+                lwmac_set_timeout(&lwmac, TIMEOUT_WAIT_FOR_DEST_WAKEUP, time_until_tx);
+
+                /* Register neighbour to be the next */
+                lwmac.tx.current_neighbour = neighbour;
+
+                /* Stop dutycycling, we're preparing to send. This prevents the
+                 * timeout arriving late, so that the destination phase would
+                 * be missed. */
```

Comment:
@daniel-k Thanks so much for this detailed reply! It is much more clear to me now! :smiley:

Besides, sorry for bothering again, I still have three minor issues to be cleared:
1) In RIOT, if there are two threads with different priorities, like LwMAC (with higher priority) and the idle thread (with lower priority), the execution pattern of RIOT should be:
pattern-a) run the higher priority thread to complete and only run the low priority thread when the higher thread is 
silent.
Or
pattern-b) run the two threads (both the high and low priority threads) at the same time with an alternative way.
I think RIOT adopt pattern-a, right? Just to be sure.

2)In case RIOT adopts execution pattern-a, assuming there are only two threads, LwMAC thread and the idle thread, and if LwMAC set up the sleep flag (```lpm_prevent_sleep &= ~(LWMAC_LPM_MASK)```) in the middle of its thread execution, the CPU will not go to sleep immediately, since LwMAC thread has higher priority than the idle thread. The CPU will only go to the sleep mode after the LwMAC thread has run to the complete of its execution and free the occupation of CPU (i.e., CPU will not go to sleep until LwMAC become silent and has freed the CPU for the idle thread to use). Does it work like this?

3) In case the CPU is now in the lpm sleep mode, a) the RTT module is independent and is still running, right? b) In case there are still only two threads,  LwMAC and the idle thread, and the CPU is now in the sleep mode, when the RTT alarm is triggered/generated, RTT callback function (i.e., ```rtt_cb()```) will be executed. But, how this RTT alarm will wakeup the CPU in the current LwMAC. Since in ```rtt_cb()```, it only sends a message to LwMAC thread which is not running, and the ```rtt_handler()``` (which sets ```lpm_prevent_sleep |= LWMAC_LPM_MASK```) is in the LwMAC thread that will only be called if LwMAC thread runs.

I'm sorry that I may ask stupid questions, (my origin major is not computer science and I also lack knowledge of OS). But, this really help me to learn more how RIOT works! Thanks! (Really hope there could be more RIOT tutorials that I can learn and dig by my self!)